File: src/2024/05/brute.kt
import java.io.File

// Global rules map
private var rules: Map<Int, List<Int>> = mapOf()

data class ParseResult(
    val rules: Map<Int, List<Int>>,
    val pages: List<List<Int>>
)

fun parseInput(): ParseResult {
    val text = File("src/2024/05/input.txt").readText()
    val (rulesStr, pagesStr) = text.split("\n\n")
    
    val pages = pagesStr.lines()
        .map { line -> line.trim()j.split(",").map { it.toInt() } }

    // Parse rules
    val rulesMap = mutableMapOf<Int, MutableList<Int>>()
    rulesStr.lines().forEach { line ->
        val (before, after) = line.split("|")
        val afterNum = after.trim().toInt()
        val beforeNum = before.trim().toInt()
        rulesMap.getOrPut(afterNum) { mutableListOf() }.add(beforeNum)
    }

    return ParseResult(rulesMap, pages)
}

fun isValid(pages: List<Int>): Boolean {
    val posLookup = pages.withIndex().associate { (index, value) -> value to index }

    for ((afterNum, beforeNums) in rules) {
        val afterPos = posLookup[afterNum] ?: continue

        for (beforeNum in beforeNums) {
            posLookup[beforeNum]?.let { beforePos ->
                if (beforePos >= afterPos) return false
            }
        }
    }
    return true
}

data class PermutationResult(
    val isValid: Boolean,
    val permutation: List<Int>
)

fun checkPermutation(perm: List<Int>): PermutationResult {
    return PermutationResult(isValid(perm), perm)
}

fun <T> List<T>.permutations(): Sequence<List<T>> = sequence {
    if (isEmpty()) yield(emptyList())
    else {
        for (i in indices) {
            val element = get(i)
            val rest = toMutableList().apply { removeAt(i) }
            rest.permutations().forEach { subPerm ->
                yield(listOf(element) + subPerm)
            }
        }
    }
}

fun part1() {
    val parseResult = parseInput()
    rules = parseResult.rules
    val result = mutableListOf<List<Int>>()
    println("Processing examples")

    parseResult.pages.forEach { pages ->
        println("Processing permutation of size ${pages.size}")
        
        // Find first valid permutation
        val validPerm = pages.permutations()
            .asSequence()
            .map { checkPermutation(it) }
            .find { it.isValid }
            ?.permutation
        
        validPerm?.let { result.add(it) }
    }

    // Add middle numbers
    val total = result.sumOf { perm ->
        perm[perm.size / 2]
    }

    println(total)
}

fun day5() {
    part1()
}

File: src/2024/05/brute.py
import os
from itertools import permutations
from math import factorial
from multiprocessing import Pool

from p_tqdm import p_uimap
from tqdm import tqdm

rules = {}


def parse_input() -> tuple[dict[int, list[int]], list[list[int]]]:
    with open(f"{os.getcwd()}/src/2024/05/input.txt") as f:
        text = f.read()

    rules_str, pages_str = text.split("\n\n")
    pages = [[int(page) for page in line.split(",")] for line in pages_str.splitlines()]

    # parse rules
    rules: dict[int, list[int]] = {}
    for line in rules_str.splitlines():
        b, a = line.split("|")
        rules[int(a)] = rules.get(int(a), []) + [int(b)]

    return rules, pages


# dict is {after: before}


def isValid(pages: list[int]) -> bool:
    global rules
    pos_lookup = {num: i for i, num in enumerate(pages)}

    for after_num, before_nums in rules.items():
        after_pos = pos_lookup.get(after_num, -1)
        if after_pos == -1:  # if after_num isn't in the sequence, skip
            continue

        for before_num in before_nums:
            if before_num in pos_lookup and pos_lookup[before_num] >= after_pos:
                return False
    return True


def check_permutation(perm) -> tuple[bool, tuple]:
    """Helper function to check a single permutation"""
    if isValid(list(perm)):
        return True, perm
    return False, perm


def part11():
    rules, examples = parse_input()
    result = []
    print("Processing examples")

    for pages in tqdm(examples):
        # Process permutations in parallel
        for valid, perm in p_uimap(check_permutation, permutations(pages), total=factorial(len(pages)), leave=False):
            if valid:
                result.append(perm)
                break  # We only need the first valid permutation

    # add middle numbers
    total = 0
    for example in result:
        total += example[len(example) // 2]

    print(total)


def part1():
    rules, examples = parse_input()
    result = []
    print("Processing examples")

    # Create a process pool
    with Pool() as pool:
        print("Processing permutations")
        for pages in tqdm(examples):
            # Prepare arguments for each permutation
            perms = permutations(pages)
            args = [(rules, perm) for perm in perms]

            # Process permutations in parallel
            for valid, perm in pool.imap_unordered(check_permutation, args):
                if valid:
                    result.append(perm)
                    break  # We only need the first valid permutation

    # add middle numbers
    total = 0
    for example in result:
        total += example[len(example) // 2]

    print(total)


if __name__ == "__main__":
    from multiprocessing import freeze_support

    freeze_support()
    part11()
